### PR TITLE
Refactor withStyles

### PR DESCRIPTION
**Pull request by behnammodi**
@kof I did but I need your help,
Items:
- Flow announcements 
- 2  Tests failed
`should render a number of composed style rules with withStyles API`
`allows nested ThemeProviders with custom namespace`

SUMMARY of Test:
`713 tests completed`
`12 tests skipped`
`2 tests failed`

### CODE REVIEW

**Comment by kof** [comment]
One of the errors from tests is the missing `props.theming` handling 

```
  React-JSS: theming withStyles()
    when theming object returned from createTheming is provided to injectSheet options
      ✖ allows nested ThemeProviders with custom namespace
        Chrome 90.0.4430 (Mac OS X 10.15.7)
      Error: expected undefined to sort of equal { color: '#aaa' }
          at Assertion.101.Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/withStylesTheming.js:197:13)
          at Assertion.101.Assertion.eql (webpack://jss/node_modules/expect.js/index.js:230:0 <- packages/react-jss/tests/withStylesTheming.js:331:10)
          at Context.<anonymous> (webpack://jss/packages/react-jss/tests/withStylesTheming.js:400:43 <- packages/react-jss/tests/withStylesTheming.js:20995:86)

```

**Comment by behnammodi** [comment]
> One of the errors from tests is the missing `props.theming` handling
> 
> ```
>   React-JSS: theming withStyles()
>     when theming object returned from createTheming is provided to injectSheet options
>       ✖ allows nested ThemeProviders with custom namespace
>         Chrome 90.0.4430 (Mac OS X 10.15.7)
>       Error: expected undefined to sort of equal { color: '#aaa' }
>           at Assertion.101.Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/withStylesTheming.js:197:13)
>           at Assertion.101.Assertion.eql (webpack://jss/node_modules/expect.js/index.js:230:0 <- packages/react-jss/tests/withStylesTheming.js:331:10)
>           at Context.<anonymous> (webpack://jss/packages/react-jss/tests/withStylesTheming.js:400:43 <- packages/react-jss/tests/withStylesTheming.js:20995:86)
> ```

Could you advise?

**Comment by kof** [comment]
when props.theming is received you need to handle that and use it instead of the theming from context, see the test

**Comment by kof** [comment]
Just looked at it more closely, I was wrong about props, we don't receive theming over props, only theme, so its in options. Your mistake was not using theming.context

**Comment by behnammodi** [comment]
> Just looked at it more closely, I was wrong about props, we don't receive theming over props, only theme, so its in options. Your mistake was not using theming.context

Yeas, I add `theming.context` but it not work again

**Comment by kof** [comment]
> Yeas, I add theming.context but it not work again

There is another bug somewhere, it passed the first 2 assertions

```
      expect(themeReceivedInComponentA).to.eql(themeA)
      expect(themeReceivedInComponentB).to.eql(themeB)
```

Creating a PR

**Comment by kof** [comment]
https://github.com/cssinjs/jss/pull/1508

see last commit

**Comment by behnammodi** [comment]
I fixed

**Comment by kof** [comment]
I think you have permissions to commit directly into that new PR

**Comment by behnammodi** [comment]
> I think you have permissions to commit directly into that new PR

I haven't permissions, but anyway I fixed on this PR

**Comment by behnammodi** [comment]
> I think you have permissions to commit directly into that new PR

Also, I apply your changes

**Comment by kof** [comment]
I changed the base of this PR so that I can commit somewhere and we can merge it. Please rebase this PR.

In the last commit I fixed the second mistake, createUseStyles() is accepting theming, not theme. This has fixed that  failing test now.

**Comment by behnammodi** [comment]
> I changed the target branch of this PR so that I can commit somewhere and we can merge it. Please rebase this PR.
> 
> In the last commit I fixed the second mistake, createUseStyles() is accepting theming, not theme. This has fixed that failing test now.

I fixed too

**Comment by behnammodi** [comment]
@kof We just have 1 error in test cases

`should render a number of composed style rules with withStyles API`

**Comment by kof** [comment]
> @kof We just have 1 error in test cases

yeah, I am confident you can figure it out

**Comment by behnammodi** [comment]
@kof Thanks man, I hope :)

**Comment by behnammodi** [comment]
@kof We don't have any corresponding test case about `should render a number of composed style rules with styled API` for `useStyles` so I'm completely confusing. I find a piece of code that is not inside `createUseStyles` but we have it on previous `withStyles`:

```js
if (sheet && registry) {
  registry.add(sheet)
}
```

And I don't know we need it or no

**Comment by kof** [comment]
@behnammodi not sure what you are confused about, from what I can see we have the register.add() call inside manageSheet()

**Comment by behnammodi** [comment]
> @behnammodi not sure what you are confused about, from what I can see we have the register.add() call inside manageSheet()

Yes, you're right 👍🏻

**Comment by kof** [comment]
@behnammodi  I am assuming you will let me know if you are completely stuck with this

**Comment by behnammodi** [comment]
> @behnammodi I am assuming you will let me know if you are completely stuck with this

Yes, In the last test case  I'm completely stuck

**Comment by kof** [comment]
ok, going to look into it

**Comment by behnammodi** [comment]
> ok, going to look into it

Awesome 👍🏻

**Comment by kof** [comment]
Fixed, please rebase

**Comment by kof** [comment]
A bunch of type errors, otherwise all tests passed

**Comment by behnammodi** [comment]
> Fixed, please rebase

Thanks man 👍🏻 🙏🏻

**Comment by behnammodi** [comment]
@kof Hi man, I fixed temporarily `flow` issue, so you can review it again

**Comment by behnammodi** [comment]
@kof sorry for late, I had very busy week 🙏🏻

File: packages/react-jss/src/withStyles.js [hunk 1]
```diff
@@ -44,23 +35,24 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
         classesProp ? mergeClasses(sheetClasses, classesProp) : sheetClasses
     )
 
+    const useStyles = createUseStyles(
+      (styles: any),
```
Comment (kof) [inline packages/react-jss/src/withStyles.js:39]: we shouldn't be using any type

File: packages/react-jss/src/withStyles.js [hunk 2]
```diff
@@ -44,23 +35,24 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
...
+      ({
+        theming,
+        index,
+        name: displayName,
+        ...sheetOptions
+      }: any)
```
Comment (kof) [inline packages/react-jss/src/withStyles.js:45]: any

File: packages/react-jss/src/withStyles.js [hunk 3]
```diff
@@ -44,23 +35,24 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
...
-            ...sheetOptions
-          }),
-        [styles, theming, index, displayName, sheetOptions]
-      )
-
       const newProps = {...props}
 
-      if (injectTheme && newProps.theme == null) newProps.theme = theme
+      if (injectTheme && newProps.theme == null) {
+        ;(newProps: any).theme = theme
```
Comment (kof) [inline packages/react-jss/src/withStyles.js:54]: - any
- I think we can avoid needing semicolon first

File: packages/react-jss/src/withStyles.js [hunk 4]
```diff
@@ -70,8 +62,11 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
     })
 
     WithStyles.displayName = `WithStyles(${displayName})`
+
+    // $FlowFixMe[prop-missing]
```
Comment (kof) [inline packages/react-jss/src/withStyles.js:66]: not sure if we can but maybe there is a way to avoid this?

Comment (behnammodi) [reply packages/react-jss/src/withStyles.js:66]: I could not find

File: packages/react-jss/src/withStyles.js (hunk 3 above)
Comment (behnammodi) [reply packages/react-jss/src/withStyles.js:54]: Yes, but semicolon added by prettier and I don't know way

File: packages/react-jss/src/withStyles.js (hunk 1 above)
Comment (behnammodi) [reply packages/react-jss/src/withStyles.js]: fixed

File: packages/react-jss/src/withStyles.js (hunk 3 above)
Comment (behnammodi) [reply packages/react-jss/src/withStyles.js]: I fixed

File: packages/react-jss/src/withStyles.js (hunk 2 above)
Comment (behnammodi) [reply packages/react-jss/src/withStyles.js]: Fixed, I found a way from your codes

**Comment by kof** [comment]
merged, thank you!

**Comment by kof** [comment]
released in https://github.com/cssinjs/jss/releases/tag/v10.7.0

**Pull request merged**